internal/forms/forms.go

```diff
@@ -0,0 +1,757 @@
+	}
+	json.NewEncoder(w).Encode(GetPostedFormData(formInstanceKey.Value))
+}
+
+// similar to GetFormData, but used when posting the form-based message to the outbox
+func GetPostedFormData(key string) FormData {
+	return postedFormData[key]
+}
+
+// handles the request for viewing a form filled-in with instance values
+func GetFormTemplate(w http.ResponseWriter, r *http.Request) {
+	formPath := r.URL.Query().Get("formPath")
+	if formPath == "" {
+		http.Error(w, "formPath query param missing", http.StatusBadRequest)
+		log.Printf("formPath query param missing %s %s", r.Method, r.URL.Path)
+		return
+	}
+
+	absPathTemplate, err := findAbsPathForTemplatePath(formPath)
+	if err != nil {
+		http.Error(w, "find the full path for requested template "+formPath, http.StatusBadRequest)
+		log.Printf("find the full path for requested template %s %s: %s", r.Method, r.URL.Path, "can't open template "+formPath)
+		return
+	}
+
+	responseText, err := fillFormTemplate(absPathTemplate, "/api/form?"+r.URL.Query().Encode(), nil, make(map[string]string))
+	if err != nil {
+		http.Error(w, "can't open template "+formPath, http.StatusBadRequest)
+		log.Printf("problem filling form template file %s %s: %s", r.Method, r.URL.Path, "can't open template "+formPath)
+		return
+	}
+
+	_, err = io.WriteString(w, responseText)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		log.Printf("can't write form into response %s %s: %s", r.Method, r.URL.Path, err)
+		return
+	}
+}
+
+// returns the user-visible filename for the message attachment that holds the form instance values
+func GetXmlAttachmentNameForForm(f Form, isReply bool) string {
+	attachmentName := filepath.Base(f.ViewerURI)
+	if isReply {
+		attachmentName = filepath.Base(f.ReplyViewerURI)
+	}
+	attachmentName = strings.TrimSuffix(attachmentName, filepath.Ext(attachmentName))
+	attachmentName = "RMS_Express_Form_" + attachmentName + ".xml"
+	if len(attachmentName) > 255 {
+		attachmentName = strings.TrimPrefix(attachmentName, "RMS_Express_Form_")
+	}
+	return attachmentName
+}
+
+// given the contents of a form attachment, finds the associated form and returns the filled-in form in HTML
+func RenderForm(contentData []byte, composereply bool) (string, error) {
+	buf := bytes.NewBuffer(contentData)
+
+	type Node struct {
+		XMLName xml.Name
+		Content []byte `xml:",innerxml"`
+		Nodes   []Node `xml:",any"`
+	}
+
+	var n1 Node
+	formParams := make(map[string]string)
+	formVars := make(map[string]string)
+
+	err := xml.NewDecoder(buf).Decode(&n1)
+	if err != nil {
+		return "", err
+	}
+
+	if n1.XMLName.Local != "RMS_Express_Form" {
+		return "", errors.New("missing RMS_Express_Form tag in form XML")
+	}
+	for _, n2 := range n1.Nodes {
+		switch n2.XMLName.Local {
+		case "form_parameters":
+			for _, n3 := range n2.Nodes {
+				formParams[n3.XMLName.Local] = string(n3.Content)
+			}
+		case "variables":
+			for _, n3 := range n2.Nodes {
+				formVars[n3.XMLName.Local] = string(n3.Content)
+			}
```

Comment:
I'd say it's a reasonable assumption. Let's say the attachment in the b2f message was utf-16 then the message would have been rejected a few lines above at: `if n1.XMLName.Local != "RMS_Express_Form" `
It is possible that some other Winlink client would produce b2f with utf-16 XML attachments, but I know that Express and my code in Pat produces utf-8. I see no need to be more flexible than as-is.

Reply:
I can see why you would think that. But making any assumptions when it comes to Winlink and character encoding has proved to be a mistake in the past. Winlink Express and CMS uses both CP-1252 and UTF-8 randomly throughout the system, even in parts of the protocol where they clearly document that ASCII is the only valid character set. 

I quick peek at the forms directory reveals this list of different character sets currently in use:
```
 184  ASCII text
  41  UTF-8 Unicode text
  13  UTF-8 Unicode (with BOM) text
  10  Non-ISO extended-ASCII text
   1  ISO-8859 text
 ```

 The majority is ASCII, which is fine since it is a subset of UTF-8. But there is still 24 files with character encoding incompatible with a pure `[]byte` to `string` cast, that would probably not be catched by the `if n1.XMLName.Local != "RMS_Express_Form"` test.

 I'd rather we at the very least warn people when using those incompatible files, since we can not guarantee that they will function properly with this implementation. We could also [strip the BOM](https://github.com/dimchansky/utfbom) to bring that number down to 11.

It is very easy to test that some arbitrary data is valid UTF-8 encoded text with [`func utf8.Valid(p []byte) bool`](https://golang.org/pkg/unicode/utf8/#Valid). Maybe we can do it for the complete template data and received files? Then we could provide a warning to the user in these cases.

Reply:
see 63904d19147e1c20ad26de62cf7f6a6d93c2a26b and 2cb14497dae8fed9b0d92654fbbe0abfc44e7b16. It's not practical to reject template html files that fails the utf8.Valid check. Instead I'm just logging a warning.

Unfortunately the popular 'ICS USA Forms/ICS213_Initial.html' contains 0x94 around line 620 in function RWS(str). They are replacing fancy quotes with normal quotes.